### PR TITLE
adds a raw data viewer for advanced users

### PR DESCRIPTION
**Pull request by didimitrie**
Changelog: 
- [x] adds a raw data viewer, similar to FE1's - but a bit more "raw" - ie, no cosmetics are applied on property names or values anywhere
- [x] adds "open in new window" for object selection

TODOs: 
- [x] fix array key indexing for object properties
- [x] fix null props (string vs. array)
- [x] fix loaded resources tabs for objects loaded only scenario
- [x] fix raw data viewer for objects loaded only scenario

Screen caps:

image

image

image

### CODE REVIEW

**Comment by benjaminvo** [comment]
Really nice @didimitrie! I think we're close to something we can release but I have some suggestions for how we can improve it both now + in a follow up PR.

**Do now**
- Keep Scene Explorer title bar in both states. 
  - Dev mode toggle button has on/off state
  - Unfold/collapse buttons can be disabled in dev mode if they do nothing
- Remove the big blue button. Rely on the button in the title bar to turn dev mode on and off.

**Follow up**
- Reuse the scene explorer's tree view in dev mode, so we give the feeling that it's the same component but you're just showing/hiding data. The styling can be more compressed in dev mode.

@andrewwallacespeckle Happy to collaborate with you on this today

CleanShot 2024-05-30 at 06 33 52@2x

**Comment by didimitrie** [comment]
> Really nice @didimitrie! I think we're close to something we can release but I have some suggestions for how we can improve it both now + in a follow up PR.
> 
> **Do now**
> - Keep Scene Explorer title bar in both states. 
>   - Dev mode toggle button has on/off state
>   - Unfold/collapse buttons can be disabled in dev mode if they do nothing
> - Remove the big blue button. Rely on the button in the title bar to turn dev mode on and off.
> 
> **Follow up**
> - Reuse the scene explorer's tree view in dev mode, so we give the feeling that it's the same component but you're just showing/hiding data. The styling can be more compressed in dev mode.
> 
> @andrewwallacespeckle Happy to collaborate with you on this today
> 
> CleanShot 2024-05-30 at 06 33 52@2x
> 

I should have some time today to implement this.  Let's ping each other if you or @andrewwallacespeckle get to it first!

File: packages/frontend-2/components/viewer/dataviewer/Row.vue [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      @click="handleExpand"
+    >
+      <div
+        class="col-span-1 truncate mr-1 flex items-center text-foreground-2 font-semibold"
+      >
+        <ChevronRightIcon
+          v-if="expandable"
+          :class="`w-3 ${expanded ? 'rotate-90' : ''} transition `"
+        />
+        <span class="select-all">{{ prop.key }}</span>
```
Comment (andrewwallacespeckle) [inline packages/frontend-2/components/viewer/dataviewer/Row.vue:21]: Nice use of `select-all`

Comment (andrewwallacespeckle) [inline packages/frontend-2/components/viewer/dataviewer/Row.vue:21]:
```suggestion
      <div
        class="col-span-1 mr-1 flex items-center text-foreground-2 font-semibold"
      >
        <ChevronRightIcon
          v-if="expandable"
          :class="`w-3 ${expanded ? 'rotate-90' : ''} transition `"
        />
        <span class="select-all truncate">{{ prop.key }}</span>
```
Currently, the first column is not truncating as expected.
image

File: packages/frontend-2/components/viewer/resources/ObjectCard.vue [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          <div :class="`text-sm sm:text-base font-bold truncate min-w-0`">
+            Object {{ object.objectId }}
+          </div>
+        </div>
+      </div>
+    </div>
+    <Transition>
+      <div
+        v-if="showRemove"
+        class="to-foundation group absolute inset-0 z-[21] flex h-full w-full items-center justify-end space-x-2 rounded bg-gradient-to-r from-blue-500/0 p-4"
```
Comment (andrewwallacespeckle) [inline packages/frontend-2/components/viewer/resources/ObjectCard.vue:22]:
```suggestion
        class="group absolute inset-0 z-[21] flex h-full w-full items-center justify-end space-x-2 rounded bg-gradient-to-r from-transparent to-foundation p-4"
```
Reordered and replaced `bg-blue-500/0` for better readability

File: packages/frontend-2/components/viewer/selection/Sidebar.vue [hunk 3]
```diff
@@ -28,6 +28,17 @@
             Isolate
           </div>
         </FormButton>
+        <div class="w-full text-right">
+          <FormButton
```
Comment (andrewwallacespeckle) [inline packages/frontend-2/components/viewer/selection/Sidebar.vue:32]: Maybe a tooltip here? @benjaminvo

**Review by andrewwallacespeckle — CHANGES_REQUESTED** [review]
A few minor suggestions, but great work!

**Comment by andrewwallacespeckle** [comment]
@benjaminvo is keen to get this out this week, so I am going to start making some these changes. I'll be at my desk in Gather if anyone wants to help

File: packages/frontend-2/components/viewer/Controls.vue [hunk 4]
```diff
@@ -201,16 +203,28 @@
...
+
+      <div v-show="resourceItems.length !== 0 && activeControl === 'jsonViewer'">
+        <!-- <KeepAlive> -->
+        <ViewerDataviewerPanel
+          class="pointer-events-auto"
+          @close="activeControl = 'none'"
+        />
+        <!-- </KeepAlive> -->
+      </div>
+
```
Comment (andrewwallacespeckle) [inline packages/frontend-2/components/viewer/Controls.vue:221]: Not needed now its in Scene Explorer

**Comment by andrewwallacespeckle** [comment]
Leaving questions here to discuss:

image
What should the tooltip be? Are we happy with Raw Data Viewer, or is Raw Data Mode maybe a better fit with the new changes?

**Comment by andrewwallacespeckle** [comment]
> @benjaminvo is keen to get this out this week, so I am going to start making some these changes. I'll be at my desk in Gather if anyone wants to help

I've made the changes suggested by Benjamin, and left a few suggestions. Will wait for Dim before going any further

**Comment by benjaminvo** [comment]
I just updated the styling and copy a bit. I've called it Dev Mode but let me know if you don't like that! I'm actually not really sure I like it.

CleanShot 2024-05-30 at 13 25 51@2x

File: packages/frontend-2/components/viewer/selection/Sidebar.vue (hunk 3 above)
Comment (benjaminvo) [reply packages/frontend-2/components/viewer/selection/Sidebar.vue:32]: Sorry I didn't see this comment. Yes I agree! But let's just follow up with that and merge this now so we don't need to wait for the CI checks again.

**Pull request merged**